src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -0,0 +1,597 @@
+        // we assume that the student may perform slightly worse that previously and dampen the prior confidence for the prediction process
+        priorConfidence *= 0.75;
+        final var scores = participantScoreService.getStudentAndTeamParticipationScoresAsDoubleStream(learningPath.getUser(), competency.getExercises()).summaryStatistics();
+        double LU = competency.getLectureUnits().size();
+        double EX = competency.getExercises().size();
+        double LO = LU + EX;
+        double MT = competency.getMasteryThreshold();
+        double EXcomp = EX - numberOfPendingExercises;
+        double a = 100d / (3d * LO);
+        double b = 100d * (LU + EXcomp + scores.getCount()) / (3d * LO) + 2d * priorConfidence / 3d - MT;
+        double c = 100d * (LU + EXcomp) * scores.getCount() / (3d * LO) + 2d * scores.getSum() / 3d - MT * scores.getCount();
+        double D = Math.sqrt(Math.pow(b, 2) - 4 * a * c);
+        double prediction1 = Math.ceil((-b + D) / (2d * a));
+        double prediction2 = Math.ceil((-b - D) / (2d * a));
+        int prediction = (int) Math.max(prediction1, prediction2);
+        // numerical edge case, can't happen for valid competencies
+        return Math.max(prediction, 0);
+    }
+
+    /**
+     * Generates a map from difficulty level to a set of corresponding exercises.
+     *
+     * @param exercises the exercises that should be contained in the map
+     * @return a map from difficulty level to a set of corresponding exercises
+     */
+    private static HashMap<DifficultyLevel, Set<Exercise>> generateDifficultyLevelMap(Set<Exercise> exercises) {
+        HashMap<DifficultyLevel, Set<Exercise>> difficultyLevelMap = new HashMap<>();
+        for (var difficulty : DifficultyLevel.values()) {
+            difficultyLevelMap.put(difficulty, new HashSet<>());
+        }
+
+        exercises.forEach(exercise -> {
+            var difficulty = exercise.getDifficulty();
+            // if no difficulty is set, assume medium difficulty level
+            if (difficulty == null) {
+                difficulty = DifficultyLevel.MEDIUM;
+            }
+            difficultyLevelMap.get(difficulty).add(exercise);
+        });
+
+        return difficultyLevelMap;
+    }
+
+    /**
+     * Computes the recommended amount of exercises per difficulty level.
+     *
+     * @param numberOfExercisesRequiredToMaster the minimum number of exercises that should be recommended
+     * @param priorConfidence                   the average confidence of all prior competencies
+     * @return array containing the recommended number of exercises per difficulty level (easy to hard)
+     */
+    private static int[] getRecommendedExerciseDistribution(int numberOfExercisesRequiredToMaster, double priorConfidence) {
+        final var distribution = getExerciseDifficultyDistribution(priorConfidence);
+        final var numberOfExercises = new int[3];
```

Comment:
```suggestion
        final var numberOfExercises = new int[DifficultyLevel.values().length];
```
Shouldn't matter, but it's clearer where the 3 comes from and one less place to potentially change in the future